Packages/com.unity.inputsystem/InputSystem/Editor/AssetImporter/InputActionImporter.cs

```diff
@@ -158,48 +158,61 @@ public override void OnImportAsset(AssetImportContext ctx)
-                    // asset import. So instead we register a callback to trigger a delayed asset refresh which should then pick up the
-                    // changed/added script, and trigger a new import.
-                    EditorApplication.delayCall += AssetDatabase.Refresh;
+                    var wrapperFilePath = m_WrapperCodePath;
+                    if (string.IsNullOrEmpty(wrapperFilePath))
+                    {
+                        // Placed next to .inputactions file.
+                        var assetPath = ctx.assetPath;
+                        var directory = Path.GetDirectoryName(assetPath);
+                        var fileName = Path.GetFileNameWithoutExtension(assetPath);
+                        wrapperFilePath = Path.Combine(directory, fileName) + ".cs";
+                    }
+                    else if (wrapperFilePath.StartsWith("./") || wrapperFilePath.StartsWith(".\\") ||
+                             wrapperFilePath.StartsWith("../") || wrapperFilePath.StartsWith("..\\"))
+                    {
+                        // User-specified file relative to location of .inputactions file.
+                        var assetPath = ctx.assetPath;
+                        var directory = Path.GetDirectoryName(assetPath);
+                        wrapperFilePath = Path.Combine(directory, wrapperFilePath);
+                    }
+                    else if (!wrapperFilePath.ToLower().StartsWith("assets/") &&
+                             !wrapperFilePath.ToLower().StartsWith("assets\\"))
+                    {
+                        // User-specified file in Assets/ folder.
+                        wrapperFilePath = Path.Combine("Assets", wrapperFilePath);
+                    }
+
+                    var dir = Path.GetDirectoryName(wrapperFilePath);
+                    if (!Directory.Exists(dir))
+                        Directory.CreateDirectory(dir);
+
+                    var options = new InputActionCodeGenerator.Options
+                    {
+                        sourceAssetPath = ctx.assetPath,
+                        namespaceName = m_WrapperCodeNamespace,
+                        className = m_WrapperClassName,
+                    };
+
+                    if (InputActionCodeGenerator.GenerateWrapperCode(wrapperFilePath, asset, options))
+                    {
+                        // When we generate the wrapper code cs file during asset import, we cannot call ImportAsset on that directly because
+                        // script assets have to be imported before all other assets, and are not allowed to be added to the import queue during
+                        // asset import. So instead we register a callback to trigger a delayed asset refresh which should then pick up the
+                        // changed/added script, and trigger a new import.
+                        EditorApplication.delayCall += AssetDatabase.Refresh;
```

Comment:
Should we not just call `AssetDatabase.ImportAsset` instead of Refresh?

Reply:
I think that should be fine. I'll give it a go. Guess there's something to Refresh being more robust in this case (by the time we run the delayed call, the path we have need no longer be valid) but I'm not sure that matters in actual practice.